### ISSUE TITLE
API interaction system adjustment

### ISSUE THREAD

**Issue by nl0** [quiltdata/quilt#684]
# The Problem

Our current API interaction system is too verbose and not very consistent.
Basically, we just call `fetch` and pass all the options manually, which is tedious and error-prone.
Also, we need to intercept 401s and log user out automatically.

# Possible Solution

A possible solution is to abstract API interaction into a "service" (injected by the provider component configured via props / context) with convenient interface (HOC that accepts some sort of request mappings / queries).

### 1. Declarative configuration
For ease of use and minimal amount of boilerplate, the configuration should be declarative, in a form of passing some sort of request mappings / queries to a HOC constructor.      

### 2. Consistency
We need a consistent way to make requests and handle responses / errors.

### 3. Centralised request / response handling
We need a way to hook into request and response flow in a single place to inject custom processing (auth hooks, json parsing, header injection, etc).

## Implementation options

We have, basically, 3 different options:

1. a wrapper around [react-refetch](https://github.com/heroku/react-refetch);

1. homegrown redux service / provider for abstracting the calls in some way;

1. GraphQL / [Apollo](https://www.apollographql.com/)

### 1. `react-refetch`

Quote:
> A simple, declarative, and composable way to fetch data for React components

#### Example implementation

API connector (provider component and consumer HOC):
```js
export const Provider = composeComponent('APIProvider',
  // get config from props and put it into context
);

export default (mappings) => composeHOC('withAPI',
  // ... inject config from context
  refetch.defaults({
    buildRequest: ({ url, headers, ...options }) => {
      // get config from props
      const { base } = ...;
      // get auth headers
      const authHeaders = ...;
      return new Request(urlJoin(base, url), { headers: { ...authHeaders, headers }, ...options });
    },
    handleResponse: async (resp)  => {
      // handle authentication loss
      if (resp.status === 401) {
        // access `dispatch` in some way
        dispatch(authLost());
        throw new NotAuthenticated();
      }
      // maybe some json processing and another stuff from current `utils/request`
      return resp;
    },
  })(mappings),
);
```

Usage (`Provider` should be present somewhere up the hierarchy when rendering this component):
```js
composeComponent('Profile',
  // declarative config, can be factored out if necessary
  withAPI((props) => ({
    packages:  `/profile`,
    updatePayment: () => ({
      updatePaymentResponse: {
        url: `/payment/`,
        method: 'POST',
        body: formData,
      },
    }),
  })),
  ({ packages, updatePayment, updatePaymentResponse }) => (
    <div>
      // quite convenient response state handling
      {renderPromiseState(packages, {
        pending: () => <Loading />,
        rejected: (reason) => <Error error={reason} />,
        fulfilled: (value) => <Packages packages={value} />,
      })}
      <button onClick={updatePayment}>Update payment info</button>
      {renderPromiseState(updatePaymentResponse, {
        pending: () => <p>updating...</p>,
        rejected: (reason) => <p>error updating payment info: {reason}</p>,
        fulfilled: (value) => <p>payment info updated</p>,
      })}
    </div>
  ));
```

#### Pros
* quite small investment
* familiarity (http requests look more familiar than QraphQL queries)

#### Cons
* the library API seems a little inflexible, so it'll require some workarounds to wire things up properly
* requests state is kept outside of redux (doesn't seem like a serious problem, but our architecture is based upon redux ecosystem, so it would be nice to use it whenever possible)
* the data is not structured / normalized properly

#### Implementation

We'll need a provider that runs the saga that issues requests and handles responses (with reducer that holds the requests state).

Also, we'll need a consumer HOC which
- selects proper slice of state and maps it to props based on config
- binds component lifecycle to request (re)fetching
- binds request functions to component props and arguments

Basically, we'd write our own opinionated version of `redux-refetch` or something similar.

#### Pros
* we can do anything we want in any way we want
* everything will be within our redux-based framework

#### Cons
* large investment

Doesn't seem like a viable option given there are established feature-rich 3rd-party tools that can be used for this.

### 3. GraphQL / Apollo

[GraphQL](https://graphql.org/learn/) is a model querying language and [Apollo](https://www.apollographql.com/) is a set of technologies for working with GraphQL all across the stack (includes client and server tools).

#### Pros
* result is a properly structured and documented API with consistent type system and data normalization capabilities
* Apollo stack is mature and very flexible
* Apollo client uses redux, so its state and transitions are more or less transparent

#### Cons
* quite large investment
* unfamiliarity (compared to http requests)
* more complex than the other options

#### Implementation

To implement a GraphQL API we need several things:

1. a GraphQL schema of the API;
1. a configured Apollo provider that connects the client to the API;
1. processing GraphQL language (queries and possibly schema) by our webpack setup;
1. a GraphQL API itself, implemented as one of these:
   1. a client-side executable schema (considered quite heavyweight, the query execution environment adds ~200kB to the bundle size);
   1. an API gateway server (probably using nodejs), which issues requests to the current REST API under the hood;
   1. a GraphQL endpoint in the current flask-based registry server;

Connecting component to data looks like this:
```js
// import query from a file
import query from './query.graphql';
// or use inline query
const query = gql`
  query Component__data {
    package { owner name }
  }
`;

composeComponent('ComponentWithData',
  graphql(query),
  ({ data }) => (
    <div>
      {renderData(data, {
        loading: () => <p>loading</p>,
        error: (error) => <p>error: {error}</p>,
        data: ({ package: { name, owner } }) => <p>package: {owner}/{name}</p>,
      })}
    </div>
  ));
```

##### 1. Schema

Looks like this:
```graphql
# quilt.graphql

type Package {
  # ...
}

# other types

type Query {
  package(owner: String!, name: String!): Package
  packages(query: String): [Package!]!
  # ...
}

type Mutation {
  deletePackage(owner: String!, name: String!): DeletePackageResult!
  # ...
  signIn(email: String!, password: String!): SignInResult!
  signUp(...): SignUpResult!
}
```

##### 2. Apollo provider

Nothing unusual, just configure an Apollo client and inject it using `ApolloProvider`.

##### 3. GraphQL processing

We'd need to use GraphQL loader for webpack to be able to store queries and fragments in separate files.

###### 4.1. Client-side executable schema

To implement executable schema on the client-side, we'd need to provide a resolver map looking like this:

```js
{
  Query: {
    package: (root, params, context) => {
      // make a request and process the data
    },
    packages: (root, params, context) => {
      // make a request and process the data
    },
    // ...
}
```

The execution environment is quite heavyweight, so this solution should be considered temporary.
The resolvers logic can be easily moved to the server-side when implementing nodejs-based GraphQL API gateway.

**Pros**:
* smaller investment than a server-side API gateway: doesn't require backend changes and new infrastructure for running the gateway server

**Cons**:
* considerable runtime overhead
* GraphQL API is not exposed, so it cannot be consumed by arbitrary GraphQL client

###### 4.2. GraphQL gateway over existing REST API

This one is quite straightforward. Suggested stack: `koa` or `express` + `apollo-server` / `apollo-tools`. The implementation is similar to the client-side solution (writing resolvers for schema that fetch and transform data), but the facade is a separate server that listens for GraphQL requests (via http or ws), makes http requests to the REST API and then responds with collected and transformed data.

**Pros**:
* better performance and smaller overhead compared to the client-side solution
* exposed API can be consumed by any GraphQL client
* better DX: something like graphiql can be used to explore the API

**Cons**:
* larger investment (infrastructure adjustment required to run a new server)
* overhead is larger compared to the native solution (see next section)

###### 4.3. GraphQL endpoint in the current flask-based registry server

Python ecosystem has decent GraphQL support, so we can serve the GraphQL stuff right from our flask app, querying the data directly using SQLA and avoiding unnecessary overhead related to REST requests.

**Pros**:
* best performance and least overhead
* smaller investment than the previous option (infrastructure adjustments not required, just add a new endpoint(s) to the existing backend)
* all the rest from the previous option

**Cons**:
* possible performance degradation of the current registry functions due to added complexity of GraphQL handling and possible deeply nested queries
* larger investment than the client-side option
 
# Alternative (quick-fix style) solution

Alternatively, we can adjust our request functions (`utils/request`) to inject headers and handle 401s in some way. It is quite quick and easy to implement, but it doesn't solve the problem as stated here in its entirety, just its auth handling aspect. Also, this doesn't seem too good architecture-wise. Nevertheless, it's a viable short-term solution.

**Comment by nl0** [comment]
@akarve @dimaryaz @meffij take a look please when you have some time

**Comment by nl0** [comment]
@akarve I would like to go with the quick-fix for now, and then (shortly) start implemented one of the more sophisticated solutions (I'd suggest GraphQL)

**Comment by nl0** [comment]
quick fix implemented in #701
